flatnav/util/VisitedNodesHandler.h

```diff
@@ -0,0 +1,175 @@
+    // initialize values to 0
+    _table = new uint32_t[_table_size]();
+  }
+
+  inline void prefetch(const uint32_t num) const {
+#ifdef USE_SSE
+    _mm_prefetch((char *)_table[num], _MM_HINT_T0);
+#endif
+  }
+
+  inline uint32_t getMark() const { return _mark; }
+
+  inline void insert(const uint32_t num) { set(num); }
+
+  inline void set(const uint32_t num) { _table[num] = _mark; }
+
+  inline uint32_t size() const { return _table_size; }
+
+  inline void reset(const uint32_t num) { _table[num] = _mark + 1; }
+
+  inline void clear() { _mark++; }
+
+  inline bool operator[](const uint32_t num) { return (_table[num] == _mark); }
+
+  ~VisitedNodesHandler() { delete[] _table; }
+
+  // copy constructor
+  VisitedNodesHandler(const VisitedNodesHandler &other)
+      : _table_size(other._table_size), _mark(other._mark) {
+
+    _table = new uint32_t[_table_size];
+    std::memcpy(_table, other._table, _table_size * sizeof(uint32_t));
+  }
+
+  // move constructor
+  VisitedNodesHandler(VisitedNodesHandler &&other) noexcept
+      : _table_size(other._table_size), _mark(other._mark),
+        _table(other._table) {
+    other._table = nullptr;
+    other._table_size = 0;
+    other._mark = 0;
+  }
+
+  // copy assignment
+  VisitedNodesHandler &operator=(const VisitedNodesHandler &other) {
+    if (this != &other) {
+      delete[] _table;
+      _table_size = other._table_size;
+      _mark = other._mark;
+      _table = new uint32_t[_table_size];
+      std::memcpy(_table, other._table, _table_size * sizeof(uint32_t));
+    }
+    return *this;
+  }
+
+  // move assignment
+  VisitedNodesHandler &operator=(VisitedNodesHandler &&other) noexcept {
+    _table_size = other._table_size;
+    _mark = other._mark;
+    _table = other._table;
+    other._table = NULL;
+    other._table_size = 0;
+    other._mark = 0;
+    return *this;
+  }
+};
+
+/**
+ *
+ * @brief Manages a pool of VisitedNodesHandler objects in a thread-safe manner.
+ *
+ * This class is designed to efficiently provide and manage a pool of
+ * VisitedNodesHandler instances for concurrent use in multi-threaded
+ * environments. It ensures that each handler can be used by only one thread at
+ * a time without the risk of concurrent access and modification.
+ *
+ * The class preallocates a specified number of VisitedNodesHandler objects to
+ * eliminate the overhead of dynamic allocation during runtime. It uses a mutex
+ * to synchronize access to the handler pool, ensuring that only one thread can
+ * modify the pool at any given time. This mechanism provides both thread safety
+ * and improved performance by reusing handler objects instead of continuously
+ * creating and destroying them.
+ *
+ * When a thread requires a VisitedNodesHandler, it can call
+ * `pollAvailableHandler()` to retrieve an available handler from the pool. If
+ * the pool is empty, the function will dynamically allocate a new handler to
```

Comment:
Do we want to specify a maximum pool size? The visited lists are pretty heavy in memory (100s of megabytes). I guess it's also fine to just mention that we'll need sufficient RAM for 1 list / thread.

Reply:
I think it's a good idea to specify the maximum pool size. Since in principal we never have more items in the pool than are available threads, we can cap this to the available number of cores on the system.